statemachine/statemachine.py

```diff
@@ -160,8 +173,63 @@ def allowed_events(self):
         ]
 
     def _process(self, trigger):
-        """This method will also handle execution queue"""
-        return trigger()
+        """Process event triggers.
+
+        The simplest implementation is the "synchronous" mode (``not queued``),
+        where the trigger will be run immediately and the result collected as the return.
+
+        .. note::
+
+            While processing the trigger, if others events are generated, they
+            will also be processed immediately, so a "nested" behavior happens.
+
+        If the machine is on ``queued`` mode, the event is put on a queue, and only the first
+        event will have the result collected.
+
+        .. note::
+            While processing the queue items, if others events are generated, they
+            will be processed sequentially (and not nested).
+
+        """
+        if not self.__queued:
+            # The machine is in "synchronous" mode
+            return trigger()
+
+        # The machine is in "queued" mode
+        # Add the trigger to queue and start processing in a loop.
+        self._external_queue.append(trigger)
+
+        # We make sure that only the first event enters the processing critical section,
+        # next events will only be put on the queue and processed by the same loop.
+        if self.__processing:
+            return
+
+        return self._processing_loop()
+
+    def _processing_loop(self):  # noqa: C901
+        """Execute the triggers in the queue in order until the queue is empty"""
+        self.__processing = True
+
+        # We will collect the first result as the processing result to keep backwards compatibility
+        # so we need to use a sentinel object instead of `None` because the first result may
+        # be also `None`, and on this case the `first_result` may be overridden by another result.
+        sentinel = object()
```

Comment:
That was so smart that I have no words, but I'm impressed that python doesn't have a better way to deal with that.

Reply:
I don't know a built-in alternative, but I've already seen this technique in the standard library. 

While searching for this answer, I've found [PEP-0661](https://peps.python.org/pep-0661/) that proposes a `from sentinels import Sentinel`, but it's still in draft.